### PR TITLE
build: add shading to -sources.jar to reflect the main jar which is shaded (#304)

### PR DESCRIPTION
**Pull request by aaronchung-bitquill**
### Summary

build: add shading to -sources.jar to reflect the main jar which is shaded (#304)
NOTE: marked as WIP since I've yet to finish all the testing. Creating it for early feedback.

### Description

Currently, the sources JAR is not shaded, but the main JAR is. This makes it difficult for users that are debugging using the source files in their IDEs. The purpose of this PR is to address that by shading the sources JAR as well so that they match the main JAR. 
The Gradle shadow plugin does not support shading the sources JAR. So a custom solution was instead made.

Changes:
- Added a new Gradle task `shadeSourcesJar` to shade the sources JAR
- Made the `assemble` task depend on the `shadeSourcesJar` task
- Replaced the `sourcesJar` task artifacts in the Maven publication with the `shadeSourcesJar` task artifacts

### Additional Reviewers

@karenc-bq @aaron-congo @sergiyv-improving

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -150,6 +156,44 @@ tasks.register<Jar>("cleanShadowJar") {
...
+
+    val sourceCharset: java.nio.charset.Charset = StandardCharsets.UTF_8
+    val shadedSourcesDir: java.nio.file.Path = file(tmpShadowedSourcesPath).toPath()
+
+    doFirst {
+        zipTree(tasks["sourcesJar"].outputs.files.asPath).visit {
+            val relativePath = this.relativePath.toString()
+
+            if (this.file.isDirectory) {
+                return@visit
```
Comment (aaron-congo) [inline build.gradle.kts:176]: Just for my own understanding, what is @visit? I haven't seen this before

Comment (aaronchung-bitquill) [reply build.gradle.kts:176]: In Kotlin, if we have nested functions, the `@` helps explicitly state which nested function to return from. 
Found a decent explanation here https://stackoverflow.com/questions/40160489/kotlin-whats-does-return-mean

File: build.gradle.kts [hunk 2]
```diff
@@ -423,4 +468,4 @@ tasks.withType<Test> {
         this.showStandardStreams = true
     }
     useJUnitPlatform()
-}
+}
```
[the remark this answers is not in the document]
Comment (aaronchung-bitquill) [reply build.gradle.kts]: Done!

File: build.gradle.kts [hunk 3]
```diff
@@ -150,6 +156,44 @@ tasks.register<Jar>("cleanShadowJar") {
     }
 }
 
+tasks.register<Jar>("shadeSourcesJar") {
+//    archiveClassifier.set("shaded-sources")
```
Comment (aaron-congo) [inline build.gradle.kts:160]: should we remove this comment?

File: build.gradle.kts [hunk 4]
```diff
@@ -150,6 +156,44 @@ tasks.register<Jar>("cleanShadowJar") {
...
+    archiveClassifier.set("sources")
+    dependsOn("sourcesJar")
+
+    val shadingOriginPath: String = shadingOriginPackage.replace('.', '/')
+    val shadingDestinationPath: String = shadingDestinationPackage.replace('.', '/')
+    val tmpShadowedSourcesPath = "${buildDir}/tmp/shadowedSources"
```
Comment (sergiyvamz) [inline build.gradle.kts:166]: Any chance we get some files in this folder left from a previous build?

**Review by sergiyvamz — APPROVED** [review]

File: build.gradle.kts (hunk 4 above)
Comment (aaronchung-bitquill) [reply build.gradle.kts:166]: Good catch! Definitely.
Yeah that is something I missed and will need to add.

File: build.gradle.kts (hunk 3 above)
Comment (aaronchung-bitquill) [reply build.gradle.kts]: Ah good catch. Removed!

File: build.gradle.kts (hunk 4 above)
Comment (aaronchung-bitquill) [reply build.gradle.kts:165]: Done!

**Pull request merged**